Network Security : Implement StrictOrigin and StrictOriginWhenCrossOr…

nmvk opened the pull request:
This pull request contains commit implementing initial steps for Improving Network Security project. As part of initial steps referer policy enums for strict-origin and strict-origin-when-cross-origin have been added to [hyper](https://github.com/hyperium/hyper/pull/943). Unit tests and additional logic has been added to handle these policies. Since enum changes are available on hyper version 0.9.11. We had to update hyper version to 0.9.11.

Hyper 0.9.11 depends on num_cpus 1.1.0. To avoid different version of num_cpus. We have updated rayon version from 0.4.0 to 0.4.3. Cargo.toml of util, style, geckolib, stylo component has been updated to use num_cpus version 1.1.0 instead of 0.2.2.  


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [ X] `./mach build -d` does not report any errors
- [ X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

…igin

Referer policy strict-origin and strict-origin-when-cross-origin changes have been implemented. Relevant unit test cases have been added. Enum for RefererPolicy has been added to hyper codebase and v 0.9.11 of hyper contains these changes.

This commit also contains changes related to upgrade of hyper from v0.9.10 to v0.9.11. Other dependencies changed are rayon, utils, num_cpus.

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/14059)
<!-- Reviewable:end -->

nmvk committed:
Network Security : Implement StrictOrigin and StrictOriginWhenCrossOrigin

Referer policy strict-origin and strict-origin-when-cross-origin changes have been implemented. Relevant unit test cases have been added. Enum for RefererPolicy has been added to hyper codebase and v 0.9.11 of hyper contains these changes.

This commit also contains changes related to upgrade of hyper from v0.9.10 to v0.9.11. Other dependencies changed are rayon, utils, num_cpus.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #13816) made this pull request unmergeable. Please resolve the merge conflicts.

nox commented on components/net/http_loader.rs:
```diff
@@ -437,6 +437,27 @@ fn no_referrer_when_downgrade_header(referrer_url: Url, url: Url) -> Option<Url>
+    if referrer_url.scheme() == "https" && url.scheme() != "https" {
+        return None;
+    }
+    return strip_url(referrer_url, true);
```
Nit: `return` is useless here.

mrnayak replied:
Can you please elaborate why return is not required?

Thanks,
Raghav

nox replied:
It's the last expression of the function, and the last expression in a function is the function's return value.

mrnayak replied:
Thank you,

I will change and commit the changes

nox commented on components/net/http_loader.rs:
```diff
@@ -437,6 +437,27 @@ fn no_referrer_when_downgrade_header(referrer_url: Url, url: Url) -> Option<Url>
+        if cross_origin {
+            return strip_url(referrer_url, true);
+        }
+        return strip_url(referrer_url, false);
```
Nit: `return` is useless here and doing `strip_url(referrer_url, cross_origin)` is shorter.

nox commented on components/net/http_loader.rs:
```diff
@@ -437,6 +437,27 @@ fn no_referrer_when_downgrade_header(referrer_url: Url, url: Url) -> Option<Url>
+
+/// https://w3c.github.io/webappsec-referrer-policy/#referrer-policy-strict-origin-when-cross-origin
+fn strict_origin_when_cross_origin(referrer_url: Url, url: Url) -> Option<Url> {
+    let cross_origin = referrer_url.origin() != url.origin();
```
Nit: you can do that after the `return None` shortcut.

nox commented on components/net/http_loader.rs:
```diff
@@ -437,6 +437,27 @@ fn no_referrer_when_downgrade_header(referrer_url: Url, url: Url) -> Option<Url>
+    let cross_origin = referrer_url.origin() != url.origin();
+    if referrer_url.scheme() == "https" && url.scheme() != "https" {
+        return None;
+    } else {
```
Nit: no need for a `else` here, removing it reduces indentation which is nice.

nox reviewed (changes requested):
Some nits.

nox commented:
@bors-servo try

bors-servo commented:
:lock: Merge conflict

nmvk committed:
Code review comments and upstream merge conflicts

Incorporated code review comments in components/net/http_loader.rs
Resolved merge conflicts in cargo.lock file. Updated ReferrerPolicy in
lib.rs

bors-servo commented:
:umbrella: The latest upstream changes (presumably #14054) made this pull request unmergeable. Please resolve the merge conflicts.

nmvk committed:
Remove return and resolve merge conflicts

Since last execution statement is the value which is returned, return
keyword has been removed from method strict_origin and
strict_origin_when_cross_origin.

Merge conflicts in Cargo.lock has been handled

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit e0e734f with merge 7586dd8...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/654)

highfive commented:
▶ Unexpected subtest result in /referrer-policy/strict-origin-when-cross-origin/http-rp/cross-origin/http-http/script-tag/cross-insecure.keep-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via script-tag using the http-rp\n                                 delivery method with keep-origin-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin-when-cross-origin/http-rp/cross-origin/http-http/script-tag/cross-insecure.no-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via script-tag using the http-rp\n                                 delivery method with no-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin-when-cross-origin/http-rp/cross-origin/http-http/script-tag/cross-insecure.swap-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via script-tag using the http-rp\n                                 delivery method with swap-origin-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin-when-cross-origin/http-rp/cross-origin/http-http/xhr-request/cross-insecure.keep-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via xhr-request using the http-rp\n                                 delivery method with keep-origin-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin-when-cross-origin/http-rp/cross-origin/http-http/xhr-request/cross-insecure.no-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via xhr-request using the http-rp\n                                 delivery method with no-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin-when-cross-origin/http-rp/cross-origin/http-http/xhr-request/cross-insecure.swap-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via xhr-request using the http-rp\n                                 delivery method with swap-origin-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin-when-cross-origin/http-rp/same-origin/http-http/script-tag/same-insecure.swap-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via script-tag using the http-rp\n                                 delivery method with swap-origin-redirect and when\n                                 the target request is same-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin-when-cross-origin/http-rp/same-origin/http-http/xhr-request/same-insecure.swap-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via xhr-request using the http-rp\n                                 delivery method with swap-origin-redirect and when\n                                 the target request is same-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin-when-cross-origin/meta-referrer/cross-origin/http-http/script-tag/cross-insecure.no-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via script-tag using the meta-referrer\n                                 delivery method with no-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin-when-cross-origin/meta-referrer/cross-origin/http-http/script-tag/cross-insecure.keep-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via script-tag using the meta-referrer\n                                 delivery method with keep-origin-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin-when-cross-origin/meta-referrer/cross-origin/http-http/script-tag/cross-insecure.swap-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via script-tag using the meta-referrer\n                                 delivery method with swap-origin-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin-when-cross-origin/meta-referrer/cross-origin/http-http/xhr-request/cross-insecure.keep-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via xhr-request using the meta-referrer\n                                 delivery method with keep-origin-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin-when-cross-origin/meta-referrer/cross-origin/http-http/xhr-request/cross-insecure.no-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via xhr-request using the meta-referrer\n                                 delivery method with no-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin-when-cross-origin/meta-referrer/cross-origin/http-http/xhr-request/cross-insecure.swap-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via xhr-request using the meta-referrer\n                                 delivery method with swap-origin-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin-when-cross-origin/meta-referrer/same-origin/http-http/script-tag/same-insecure.swap-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                               </span><span class="stdout">  document served over http requires an http\n                                 sub-resource via script-tag using the meta-referrer\n                                 delivery method with swap-origin-redirect and when\n                                 the target request is same-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin-when-cross-origin/meta-referrer/same-origin/http-http/xhr-request/same-insecure.swap-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via xhr-request using the meta-referrer\n                                 delivery method with swap-origin-redirect and when\n                                 the target request is same-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/http-rp/cross-origin/http-http/script-tag/insecure-protocol.keep-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via script-tag using the http-rp\n                                 delivery method with keep-origin-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/http-rp/cross-origin/http-http/script-tag/insecure-protocol.no-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via script-tag using the http-rp\n                                 delivery method with no-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/http-rp/cross-origin/http-http/script-tag/insecure-protocol.swap-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via script-tag using the http-rp\n                                 delivery method with swap-origin-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/http-rp/cross-origin/http-http/xhr-request/insecure-protocol.keep-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via xhr-request using the http-rp\n                                 delivery method with keep-origin-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/http-rp/cross-origin/http-http/xhr-request/insecure-protocol.no-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via xhr-request using the http-rp\n                                 delivery method with no-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/http-rp/cross-origin/http-http/xhr-request/insecure-protocol.swap-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via xhr-request using the http-rp\n                                 delivery method with swap-origin-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/http-rp/same-origin/http-http/script-tag/insecure-protocol.keep-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via script-tag using the http-rp\n                                 delivery method with keep-origin-redirect and when\n                                 the target request is same-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/http-rp/same-origin/http-http/script-tag/insecure-protocol.no-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via script-tag using the http-rp\n                                 delivery method with no-redirect and when\n                                 the target request is same-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/http-rp/same-origin/http-http/script-tag/insecure-protocol.swap-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via script-tag using the http-rp\n                                 delivery method with swap-origin-redirect and when\n                                 the target request is same-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/http-rp/same-origin/http-http/xhr-request/insecure-protocol.no-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via xhr-request using the http-rp\n                                 delivery method with no-redirect and when\n                                 the target request is same-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/http-rp/same-origin/http-http/xhr-request/insecure-protocol.keep-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via xhr-request using the http-rp\n                                 delivery method with keep-origin-redirect and when\n                                 the target request is same-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/http-rp/same-origin/http-http/xhr-request/insecure-protocol.swap-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via xhr-request using the http-rp\n                                 delivery method with swap-origin-redirect and when\n                                 the target request is same-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/meta-referrer/cross-origin/http-http/script-tag/insecure-protocol.keep-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via script-tag using the meta-referrer\n                                 delivery method with keep-origin-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/meta-referrer/cross-origin/http-http/script-tag/insecure-protocol.no-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via script-tag using the meta-referrer\n                                 delivery method with no-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/meta-referrer/cross-origin/http-http/script-tag/insecure-protocol.swap-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via script-tag using the meta-referrer\n                                 delivery method with swap-origin-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/meta-referrer/cross-origin/http-http/xhr-request/insecure-protocol.keep-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via xhr-request using the meta-referrer\n                                 delivery method with keep-origin-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/meta-referrer/cross-origin/http-http/xhr-request/insecure-protocol.no-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via xhr-request using the meta-referrer\n                                 delivery method with no-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/meta-referrer/cross-origin/http-http/xhr-request/insecure-protocol.swap-origin-redirect.ht</span><span class="stdout">tp.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via xhr-request using the meta-referrer\n                                 delivery method with swap-origin-redirect and when\n                                 the target request is cross-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/meta-referrer/same-origin/http-http/script-tag/insecure-protocol.keep-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via script-tag using the meta-referrer\n                                 delivery method with keep-origin-redirect and when\n                                 the target request is same-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/meta-referrer/same-origin/http-http/script-tag/insecure-protocol.no-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via script-tag using the meta-referrer\n                                 delivery method with no-redirect and when\n                                 the target request is same-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/meta-referrer/same-origin/http-http/script-tag/insecure-protocol.swap-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via script-tag using the meta-referrer\n                                 delivery method with swap-origin-redirect and when\n                                 the target request is same-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/meta-referrer/same-origin/http-http/xhr-request/insecure-protocol.keep-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via xhr-request using the meta-referrer\n                                 delivery method with keep-origin-redirect and when\n                                 the target request is same-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/meta-referrer/same-origin/http-http/xhr-request/insecure-protocol.no-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via xhr-request using the meta-referrer\n                                 delivery method with no-redirect and when\n                                 the target request is same-origin.
    
      ▶ Unexpected subtest result in /referrer-policy/strict-origin/meta-referrer/same-origin/http-http/xhr-request/insecure-protocol.swap-origin-redirect.http.html:
      └ PASS [expected FAIL] The referrer URL is origin when a\n                                 document served over http requires an http\n                                 sub-resource via xhr-request using the meta-referrer\n                                 delivery method with swap-origin-redirect and when\n                                 the target request is same-origin.
    
      ▶ TIMEOUT [expected CRASH] /_mozilla/mozilla/sigsegv.html
      │ 
      │ VMware, Inc.
      │ Gallium 0.4 on softpipe
      └ 3.3 (Core Profile) Mesa 12.0.1

jdm commented:
Look at all those newly-passing tests! The timeout is #14067 and can be ignored, but the rest need to have their [expected results](https://github.com/servo/servo/blob/master/tests/wpt/README.md#updating-test-expectations) updated.

jdm commented:
Also, it looks like there was a rebase error in cef/Cargo.lock: https://github.com/servo/servo/pull/14059/files#diff-f58e90618d14a4b6a043be78f4a4af21R2882

Darshan266 committed:
wpt-test for strict-origin and strict-origin-when-cross-origin referrer policy

Fixed rebase issue in ports/cef/Cargo.lock
Deleted PASSing testcases which were failing before implementation of referrer policies

gurudarshan266 commented:
@jdm All the .ini files corresponding to PASSing testcases have been removed for strict-origin and strict-origin-when-cross-origin referrer policies

nox commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit e0132b9 has been approved by `nox`

<!-- @bors-servo r=nox e0132b98cc223a666778ab8280b675307929c3b9 -->

bors-servo commented:
:hourglass: Testing commit e0132b9 with merge dd34b2a...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3939), [arm64](http://build.servo.org/builders/arm64/builds/3945), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4038), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/681), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/680), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4023), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/4019), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/703), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/716), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1951)
